src/game/client/gui/controlbar/controlbarcommand.cpp

```diff
@@ -125,9 +134,882 @@ const Image *ControlBar::Calculate_Veterancy_Overlay_For_Thing(const ThingTempla
+{
+    static const CommandButton *cancel_unit_command = Find_Command_Button("Command_CancelUnitCreate");
+    static const CommandButton *cancel_upgrade_command = Find_Command_Button("Command_CancelUpgradeCreate");
+    Reset_Build_Queue_Data();
+    static bool ids_initialized = false;
+    static NameKeyType build_queue_ids[QUEUE_ENTRY_COUNT];
+
+    if (!ids_initialized) {
+        Utf8String name;
+
+        for (int i = 0; i < QUEUE_ENTRY_COUNT; i++) {
+            name.Format("ControlBar.wnd:ButtonQueue%02d", i + 1);
+            build_queue_ids[i] = g_theNameKeyGenerator->Name_To_Key(name.Str());
+        }
+
+        ids_initialized = true;
+    }
+
+    for (int i = 0; i < QUEUE_ENTRY_COUNT; i++) {
+        m_queueData[i].control = g_theWindowManager->Win_Get_Window_From_Id(m_contextParent[3], build_queue_ids[i]);
+        m_queueData[i].control->Win_Enable(false);
+        m_queueData[i].control->Win_Clear_Status(WIN_STATUS_USE_OVERLAY_STATES);
+        Gadget_Button_Set_Text(m_queueData[i].control, U_CHAR(""));
+        Gadget_Button_Draw_Overlay_Image(m_queueData[i].control, nullptr);
+    }
+
+    ProductionUpdateInterface *production = producer->Get_Production_Update_Interface();
+
+    if (production != nullptr) {
+        int count = 0;
+
+        for (const ProductionEntry *entry = production->First_Production(); entry != nullptr && count < QUEUE_ENTRY_COUNT;
+             entry = production->Next_Production(entry)) {
+            if (entry->Get_Type() == ProductionEntry::PRODUCTION_UNIT) {
+                Set_Control_Command(m_queueData[count].control, cancel_unit_command);
+                m_queueData[count].type = ProductionEntry::PRODUCTION_UNIT;
+                m_queueData[count].production_id = entry->Get_Production_ID();
+                m_queueData[count].control->Win_Enable(true);
+                m_queueData[count].control->Win_Set_Status(WIN_STATUS_USE_OVERLAY_STATES);
+                Gadget_Button_Set_Enabled_Image(
+                    m_queueData[count].control, entry->Get_Production_Object()->Get_Button_Image());
+                Gadget_Button_Draw_Overlay_Image(
+                    m_queueData[count].control, Calculate_Veterancy_Overlay_For_Thing(entry->Get_Production_Object()));
+            } else {
+                Set_Control_Command(m_queueData[count].control, cancel_upgrade_command);
+                m_queueData[count].type = ProductionEntry::PRODUCTION_UPGRADE;
+                m_queueData[count].upgrade_template = entry->Get_Production_Upgrade();
+                m_queueData[count].control->Win_Enable(true);
+                m_queueData[count].control->Win_Set_Status(WIN_STATUS_USE_OVERLAY_STATES);
+                Gadget_Button_Set_Enabled_Image(
+                    m_queueData[count].control, entry->Get_Production_Upgrade()->Get_Button_Image());
+            }
+
+            count++;
+        }
+    }
+
+    m_displayedQueueCount = production->Get_Production_Count();
```

Comment:
This is dereferencing NULL. It probably needs to be one scope up.

Reply:
fixed.